### ISSUE TITLE
Memory issue when inter_op_parallelism_threads > 1 on Ubuntu 16.04

### ISSUE THREAD

**Issue by btaba** [tensorflow/tensorflow#22098]
### System information

- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes, see below
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: 
```
== cat /etc/issue ===============================================
Linux ml 4.4.0-87-generic #110-Ubuntu SMP Tue Jul 18 12:55:35 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
VERSION="16.04.3 LTS (Xenial Xerus)"
VERSION_ID="16.04"
VERSION_CODENAME=xenial

== are we in docker =============================================
No

== compiler =====================================================
c++ (Ubuntu 5.4.0-6ubuntu1~16.04.9) 5.4.0 20160609
Copyright (C) 2015 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.


== uname -a =====================================================
Linux ml 4.4.0-87-generic #110-Ubuntu SMP Tue Jul 18 12:55:35 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux

== check pips ===================================================
numpy (1.14.3)
protobuf (3.6.1)
tensorflow (1.10.0)
tensorflow-tensorboard (0.4.0)

== check for virtualenv =========================================
True

== tensorflow import ============================================
tf.VERSION = 1.10.0
tf.GIT_VERSION = v1.10.0-0-g656e7a2b34
tf.COMPILER_VERSION = v1.10.0-0-g656e7a2b34
Sanity check: array([1], dtype=int32)
/home/btabanpour/mlpy/lib/python3.5/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters

== env ==========================================================
LD_LIBRARY_PATH /usr/local/cuda/lib64:/home/anaconda3/lib/
DYLD_LIBRARY_PATH is unset

```
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**: NA
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: 1.10.0
- **Python version**: Python 3.5.2
- **Bazel version (if compiling from source)**: NA
- **GCC/Compiler version (if compiling from source)**: NA
- **CUDA/cuDNN version**: NA
- **GPU model and memory**: NA
- **Exact command to reproduce**:

Save this file as `reproduce.py`:

```python
import numpy as np
from tensorflow.keras.applications.inception_v3 import InceptionV3
from memory_profiler import profile

image = np.random.random((1, 299, 299, 3))

@profile
def predict():
    m.predict(image)


if __name__ == '__main__':
    m = InceptionV3()  # or any other large network like VGG19()
    for _ in range(100):
        predict()
```

I run `python reproduce.py` and memory increases from ~500MiB to ~1200MiB after 100 calls of `predict()`, **running on CPU**. Here is truncated output:

```
 > python reproduce.py
2018-09-05 14:41:10.229405: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
Filename: reproduce.py

Line #    Mem usage    Increment   Line Contents
================================================
     7    493.0 MiB    493.0 MiB   @profile
     8                             def predict():
     9    524.8 MiB     31.8 MiB       m.predict(image)


Filename: reproduce.py

Line #    Mem usage    Increment   Line Contents
================================================
     7    524.8 MiB    524.8 MiB   @profile
     8                             def predict():
     9    557.2 MiB     32.4 MiB       m.predict(image)


...


Filename: reproduce.py

Line #    Mem usage    Increment   Line Contents
================================================
     7   1283.4 MiB   1283.4 MiB   @profile
     8                             def predict():
     9   1283.6 MiB      0.2 MiB       m.predict(image)


Filename: reproduce.py

Line #    Mem usage    Increment   Line Contents
================================================
     7   1283.6 MiB   1283.6 MiB   @profile
     8                             def predict():
     9   1285.9 MiB      2.3 MiB       m.predict(image)

```

### Describe the problem

In short, memory keeps increasing after each predict call. I tried the above script with other networks such as `tensorflow.keras.applications.vgg19` and pure tensorflow convnets and I see the same issue.

This issue seems to be resolved when I set `inter_op_parallelism_threads=1` in the `tf.ConfigProto` as such:

```python
    config = tf.ConfigProto(
        inter_op_parallelism_threads=1)
    sess = tf.Session(config=config)
```

And re-appears when I start increasing `inter_op_parallelism_threads` to 32 (I'm running on 32 CPU cores).

### Another issue:

As I mentioned previously, setting `inter_op_parallelism_threads=1` stops the memory from increasing on each predict call. Here is a script that sets the config `inter_op_parallelism_threads=1` and does not exhibit a memory leak:
```python
import numpy as np
import tensorflow as tf
import tensorflow.keras.backend as K

from tensorflow.keras.applications.inception_v3 import InceptionV3
from memory_profiler import profile


image = np.random.random((1, 299, 299, 3))


@profile
def predict():
    m.predict(image)


if __name__ == '__main__':
    config = tf.ConfigProto(
        inter_op_parallelism_threads=1)
    sess = tf.Session(config=config)

    print(sess._config)
    K.set_session(sess)

    m = InceptionV3()
    for _ in range(100):
        predict()
```

However, I see memory increasing again if I add an extra call to `tf.Session()` before calling the one with the `tf.ConfigProto`:

```python
import numpy as np
import tensorflow as tf
import tensorflow.keras.backend as K

from tensorflow.keras.applications.inception_v3 import InceptionV3
from memory_profiler import profile

image = np.random.random((1, 299, 299, 3))


@profile
def predict():
    m.predict(image)


if __name__ == '__main__':
    s = tf.Session()  # Added an extra tf.Session() call here!!!
    s.close()

    config = tf.ConfigProto(
        inter_op_parallelism_threads=1)
    sess = tf.Session(config=config)

    print(sess._config)
    K.set_session(sess)

    m = InceptionV3()

    for _ in range(100):
        predict()
```

It seems like the ConfigProto from the first `tf.Session()` is overwriting the config from the next session I create.

---

Here is a graph of the script above for 1000 predict calls:

![tf_issue](https://user-images.githubusercontent.com/7320238/45115233-3cfa3a80-b11d-11e8-9b5e-781508fc8583.png)

**Comment by brdw** [comment]
This is a pretty big performance issue.

**Comment by btaba** [comment]
@shivaniag Any chance someone could look at this just to confirm that this is a bug?

**Comment by jiarenyf** [comment]
@tensorflow-jenkins

**Comment by tensorflowbutler** [comment]
Nagging Assignee @shivaniag: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by brdw** [comment]
bumpiddy

**Comment by KatyaRuth** [comment]
I am running into the same issue - I can set `inter_op_parallelism_threads=1` to stop the leak but that affects the latency of my model.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @shivaniag: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by celinew1221** [comment]
Same issue. With inter_opt_parallelism_threads=1, it increases slower than without the option, but still this is a big performance issue. I need the memory to do other stuff too.

**Comment by celinew1221** [comment]
Here's how I solve this issue: use placeholder - there is a make_initialize_iterator() function in dataset, use that and update your tensor in each iteration.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @shivaniag: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by Laser-Cat** [comment]
Have this problem too during training on CPU. Setting inter_op_parallelism_threads=1 as @btaba proposed gets rid of the leak. 
BTW we've never noticed this issue when training the same model with GPU.

**Comment by aashish-0393** [comment]
Facing the same issue. After around 500 predictions , memory is fully occupied. can't do any more task.

**Comment by LordSycho** [comment]
Same issue I am facing. Any update will be helpful? 
Thanks

**Comment by iamtodor** [comment]
subscription to this thread

**Comment by jiarenyf** [comment]
Met the same problem, under watching.

**Comment by shivaniag** [comment]
@fchollet  could you take a look.

**Comment by gary-harpaz** [comment]
Same problem

**Comment by markwhen** [comment]
I've met the same problem, set inter_op_parallelism_threads to 1 is ok, but the latency will increase in high load.

**Comment by markwhen** [comment]
Hi all, I find that in the same case, if we change the ConfigProto.device_count of "CPU" to 1, the mem leak disappear.
We noticed that in document, ConfigProto.device_count doesn't mean the number of cpu cores, but the number of CPU device. So in common situations, it should be 1.

**Comment by juve** [comment]
I am seeing a memory leak that looks like the same issue. I get it with anaconda::tensorflow from conda, but not with conda-forge::tensorflow nor the tensorflow from pypi.

**Comment by brandonbell11** [comment]
same issue. TF2.0, running multiple model.predict() in a loop balloons memory (GPU) and I get OOM errors.

**Comment by novog** [comment]
I can confirm that I am experiencing this issue with TF 2.0.0, as well.

**Comment by LiberiFatali** [comment]
Got this issue with TF 1.13.1

**Comment by nsarang** [comment]
Any updates?

**Comment by SergioG-M** [comment]
Still got this with TF 2.3.1

**Comment by bzamecnik** [comment]
As @btaba shows in http://blog.tabanpour.info/projects/2018/09/07/tf-docker-kube.html it's a problem of over-comitting threads to limited CPU resources, typically in docker or k8s. Both inter & intra thread count should be set to fit within the CPU limit. A detection routine is available in the article above but could be adapted to better handle for fractional CPU limits.

**Comment by sushreebarsa** [comment]
@btaba  It looks like you are using an older Version of Tensorflow (1.x) which is out of support window. Many bugs have been fixed in the latest version. Can you please execute your code using Latest Version (2.6.0) and let us know if the issue still persists? Thanks!
